src/query/storages/fuse/src/operations/merge_into/mutator/delete_by_expr_mutator.rs

```diff
@@ -16,38 +16,94 @@ use common_exception::Result;
 use common_expression::eval_function;
 use common_expression::types::BooleanType;
 use common_expression::types::DataType;
+use common_expression::BlockEntry;
 use common_expression::DataBlock;
 use common_expression::Evaluator;
 use common_expression::Expr;
 use common_expression::FunctionContext;
+use common_expression::Value;
 use common_functions::BUILTIN_FUNCTIONS;
 use common_sql::executor::cast_expr_to_non_null_boolean;
-
 pub struct DeleteByExprMutator {
     expr: Option<Expr>,
     row_id_idx: usize,
     func_ctx: FunctionContext,
+    origin_input_columns: usize,
 }
 
 impl DeleteByExprMutator {
-    pub fn create(expr: Option<Expr>, func_ctx: FunctionContext, row_id_idx: usize) -> Self {
+    pub fn create(
+        expr: Option<Expr>,
+        func_ctx: FunctionContext,
+        row_id_idx: usize,
+        origin_input_columns: usize,
+    ) -> Self {
         Self {
             expr,
             row_id_idx,
             func_ctx,
+            origin_input_columns,
         }
     }
 
-    // return block after delete, and the rowIds which are deleted
     pub fn delete_by_expr(&self, data_block: DataBlock) -> Result<(DataBlock, DataBlock)> {
+        // it's the first update, after update, we need to add a filter column
+        if data_block.num_columns() == self.origin_input_columns {
+            // assert_eq!(expr.data_type(), &DataType::Boolean);
+            self.delete_block(data_block, false)
+        } else {
+            self.delete_block(data_block, true)
+        }
+    }
+
+    fn get_filter(
+        &self,
+        data_block: &DataBlock,
+    ) -> Result<(Value<BooleanType>, Value<BooleanType>)> {
+        // filter rows that's is not processed
+        let filter_entry = data_block.get_by_offset(data_block.num_columns() - 1);
+        let old_filter: Value<BooleanType> = filter_entry.value.try_downcast().unwrap();
+        // false means this row is not processed, so use `not` to reverse it.
+        let (filter_not, _) = eval_function(
+            None,
+            "not",
+            [(old_filter.clone().upcast(), DataType::Boolean)],
+            &self.func_ctx,
+            data_block.num_rows(),
+            &BUILTIN_FUNCTIONS,
+        )?;
+        let filter_not = filter_not.try_downcast().unwrap();
+        Ok((old_filter, filter_not))
+    }
+
+    // return block after delete, and the rowIds which are deleted
+    pub fn delete_block(
```

Comment:
This function has some duplicate codes, may be we can refactor to make it more  clear.

Reply:
ok